### PR TITLE
[feat] Support parameter and variable access directly from the class body

### PR DESCRIPTION
**Pull request by jjotero**
This PR implements the attribute access to variables and parameters directly from the class body. Prior to this PR, an attempt to access either variables or parameters from the class body resulted in a `NameError` since none of these built-ins are stored in the public namespace from the class. This PR not only implements the access to these items, but also provides a more robust error handling in case of erroneous user input:

- Accessing a required (undefined) variable in the class body will raise a `ValueError`:
```python
class Foo(rfm.RegressionMixin):
    v = variable(str)
    x = f'{v}' # This will raise an error because `v` has no value
```
- Accessing a defined variable will return its default value:
```python
class Foo(rfm.RegressionMixin):
    v = variable(int, value=4)
    x = v + 1 # this sets x to 5
```
- Accessing a parameter directly from the class body is disallowed.
```python
class Foo(rfm.RegressionMixin):
    p = parameter([1, 2, 3])
    x = f'{p}' # This raises an error. Here `P` is a TestParam object, not a test parameter.
```

### CODE REVIEW

File: reframe/core/pipeline.py [hunk 1]
```diff
@@ -138,9 +138,22 @@ class RegressionMixin(metaclass=RegressionTestMeta):
...
+                not self._rfm_var_space.vars[name].is_defined()):
+                raise AttributeError(
+                    f'required variable {name!r} has not been set'
+                ) from None
+            else:
+                super().__getattr__(name)
 
 
-class RegressionTest(jsonext.JSONSerializable, metaclass=RegressionTestMeta):
+class RegressionTest(jsonext.JSONSerializable, RegressionMixin):
```
Comment (vkarak) [inline reframe/core/pipeline.py:156]: I would put `RegressionMixin` first in the row here.

File: reframe/core/meta.py [hunk 2]
```diff
@@ -28,6 +28,38 @@ def __setitem__(self, key, value):
...
+                return super().__getitem__(key)
+            except Exception as err:
+                try:
+                    # Handle variable access
+                    var = self['_rfm_local_var_space'][key]
+                    if var.is_defined():
+                        return var.default_value
+                    else:
+                        raise ValueError(
+                            f'variable {key!r} does not have an assigned value'
```
Comment (vkarak) [inline reframe/core/meta.py:48]:
```suggestion
                            f'variable {key!r} is not assigned a value'
```

File: reframe/core/meta.py [hunk 3]
```diff
@@ -28,6 +28,38 @@ def __setitem__(self, key, value):
...
+        def __getitem__(self, key):
+            '''Expose and control access to the local namespaces.
+
+            Variables may only be retrieved if their value has been previously
+            set. Accessing a parameter in the class body is disallowed (the
+            actual test parameter is set during the class instantiation).
+            '''
+            try:
+                return super().__getitem__(key)
+            except Exception as err:
```
Comment (vkarak) [inline reframe/core/meta.py:40]: Why not just a `KeyError`?

File: reframe/core/meta.py [hunk 4]
```diff
@@ -28,6 +28,38 @@ def __setitem__(self, key, value):
...
+                    # Handle parameter access
+                    if key in self['_rfm_local_param_space']:
+                        raise ValueError(
+                            'accessing a test parameter from the class '
+                            'body is disallowed'
+                        )
+                    else:
+                        # If 'key' is neither a variable nor a parameter,
+                        # raise the exception from the base __getitem__.
+                        raise err
```
Comment (vkarak) [inline reframe/core/meta.py:61]:
```suggestion
                        raise err from None
```

**Review by vkarak — CHANGES_REQUESTED** [review]

File: reframe/core/meta.py (hunk 3 above)
Comment (jjotero) [reply reframe/core/meta.py]: Yeah, makes sense. Changed.

**Review by vkarak — APPROVED** [review]
lgtm

**Review by ekouts — APPROVED** [review]

**Review by teojgo — APPROVED** [review]

**Pull request merged**